apiserver/facades/agent/uniter/networkinfoiaas.go

```diff
@@ -274,61 +281,70 @@ func (n *NetworkInfoIAAS) populateMachineNetworkInfos() error {
+	if _, ok := n.machineNetworkInfos[network.AlphaSpaceId]; !ok && spaceSet.Contains(network.AlphaSpaceId) {
 		if privateLinkLayerAddress != nil {
-			r.Info, _ = addAddressToResult(r.Info, privateLinkLayerAddress)
+			n.addAddressToResult(network.AlphaSpaceId, privateLinkLayerAddress)
 		} else {
-			r.Info = []params.NetworkInfo{{
-				Addresses: []params.InterfaceAddress{{
-					Address: privateIPAddress,
-				}},
-			}}
+			n.machineNetworkInfos[network.AlphaSpaceId] = params.NetworkInfoResult{
+				Info: []params.NetworkInfo{{Addresses: []params.InterfaceAddress{{Address: privateIPAddress}}}},
+			}
 		}
-
-		results[network.AlphaSpaceId] = r
 	}
 
 	for _, id := range spaceSet.Values() {
-		if _, ok := results[id]; !ok {
-			results[id] = params.NetworkInfoResult{
-				Error: apiservererrors.ServerError(
-					errors.Errorf("machine %q has no devices in space %q", machineID, id)),
-			}
+		if _, ok := n.machineNetworkInfos[id]; !ok {
+			n.machineNetworkInfos[id] = params.NetworkInfoResult{Error: apiservererrors.ServerError(
+				errors.Errorf("machine %q has no addresses in space %q", machineID, id))}
 		}
 	}
 
-	n.machineNetworkInfos = results
 	return nil
 }
 
-// Add address to a device in list or create a new device with this address.
-func addAddressToResult(networkInfos []params.NetworkInfo, address *state.Address) ([]params.NetworkInfo, error) {
-	deviceAddr := params.InterfaceAddress{
-		Address: address.Value(),
-		CIDR:    address.SubnetCIDR(),
-	}
-	for i := range networkInfos {
-		networkInfo := &networkInfos[i]
-		if networkInfo.InterfaceName == address.DeviceName() {
-			networkInfo.Addresses = append(networkInfo.Addresses, deviceAddr)
-			return networkInfos, nil
-		}
+// populateMachineNetworkInfoErrors populates a network
+// info result error for all of the input spaces.
+func (n *NetworkInfoIAAS) populateMachineNetworkInfoErrors(spaces set.Strings, err error) {
+	res := params.NetworkInfoResult{Error: apiservererrors.ServerError(
+		errors.Annotate(err, "getting devices addresses"))}
+	for _, id := range spaces.Values() {
+		n.machineNetworkInfos[id] = res
 	}
+}
+
+// addAddressToResult adds the network info representation
+// of the input address to the results for the input space.
+func (n *NetworkInfoIAAS) addAddressToResult(spaceID string, address *state.Address) {
+	r := n.machineNetworkInfos[spaceID]
 
 	var MAC string
 	device, err := address.Device()
 	if err == nil {
 		MAC = device.MACAddress()
 	} else if !errors.IsNotFound(err) {
-		return nil, errors.Trace(err)
+		r.Error = apiservererrors.ServerError(err)
+		n.machineNetworkInfos[spaceID] = r
+		return
+	}
+
+	deviceAddr := params.InterfaceAddress{
+		Address: address.Value(),
+		CIDR:    address.SubnetCIDR(),
+	}
+	for i := range r.Info {
+		if r.Info[i].InterfaceName == address.DeviceName() {
+			r.Info[i].Addresses = append(r.Info[i].Addresses, deviceAddr)
+			n.machineNetworkInfos[spaceID] = r
+			return
```

Comment:
I don't get why your return early here? Are you saying that you'll only ever have one address per device (sorry I don't know enough about this).

Reply:
No; if we find that we've already added the address' device, append it to the device's list. Otherwise fall through and add a new device with the address.

Reply:
Ah, now I get it, further up.